ADD: GNSS Driver for U-BLOX M10 & Support for UBX Messages

mayankmahajan-nxp opened the pull request:
MODEM_UBX: Adds Support for UBX Messages in Modem Subsystem.

GNSS API Supported: get_supported_systems, set_fix_rate, get_fix_rate,
	set_enabled_systems, get_enabled_systems, set_navigation_mode,
	get_navigation_mode.

Boards Tested: MIMXRT1062_FMURT6, VMU_RT1170.

Note: Partial support for U-BLOX Messages is provided as of now.

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>
Signed-off-by: Mayank Mahajan <mayankmahajan.x@nxp.com>

mayankmahajan-nxp committed:
ADD: GNSS Driver for U-BLOX M10 & Support for UBX Messages

MODEM_UBX: Adds Support for UBX Messages in Modem Subsystem.

GNSS API Supported: get_supported_systems, set_fix_rate, get_fix_rate,
	set_enabled_systems, get_enabled_systems, set_navigation_mode,
	get_navigation_mode.

Boards Tested: MIMXRT1062_FMURT6, VMU_RT1170.

Note: Partial support for U-BLOX Messages is provided as of now.

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>
Signed-off-by: Mayank Mahajan <mayankmahajan.x@nxp.com>

bjarki-trackunit commented on drivers/gnss/CMakeLists.txt:
```diff
@@ -9,3 +9,5 @@ zephyr_library_sources_ifdef(CONFIG_GNSS_NMEA0183 gnss_nmea0183.c)
 zephyr_library_sources_ifdef(CONFIG_GNSS_NMEA_GENERIC gnss_nmea_generic.c)
 zephyr_library_sources_ifdef(CONFIG_GNSS_QUECTEL_LCX6G gnss_quectel_lcx6g.c)
+zephyr_library_sources_ifdef(CONFIG_GNSS_U_BLOX_M10 gnss_u_blox_m10.c)
+zephyr_library_sources_ifdef(CONFIG_U_BLOX_PROTOCOL u_blox_protocol/u_blox_protocol.c)
```
This should be namespaced `GNSS_`

bjarki-trackunit commented on drivers/gnss/Kconfig:
```diff
@@ -60,11 +60,18 @@ config GNSS_INIT_PRIORITY
 
+config U_BLOX_PROTOCOL
+	bool "U-BLOX protocol"
+	default n
```
`default n` is already the default :) you should omit this line

bjarki-trackunit commented on drivers/gnss/Kconfig.u_blox_m10:
```diff
@@ -0,0 +1,30 @@
+	default y
+	depends on GNSS
+	depends on DT_HAS_U_BLOX_M10_ENABLED
+	select UART_INTERRUPT_DRIVEN
```
The UART API type is specified by the application, the driver should only select `MODEM_BACKEND_UART`

bjarki-trackunit commented on drivers/gnss/u_blox_protocol/u_blox_protocol.c:
```diff
@@ -0,0 +1,217 @@
+	460800,
+};
+
+static void u_blox_create_frame(uint8_t ubx_frame[], uint16_t *ubx_frame_size,
```
please use `uint8_t *ubx_frame` instead of `ubx_frame[]`, the latter should only be used for fixed size arrays.

bjarki-trackunit commented on drivers/gnss/u_blox_protocol/u_blox_protocol.c:
```diff
@@ -0,0 +1,217 @@
+
+static void u_blox_create_frame(uint8_t ubx_frame[], uint16_t *ubx_frame_size,
+				uint8_t message_class, uint8_t message_id,
+				uint8_t payload[], uint16_t payload_size)
```
the payload arg should be `const uint8_t *payload`

bjarki-trackunit commented on drivers/gnss/u_blox_protocol/u_blox_protocol.c:
```diff
@@ -0,0 +1,217 @@
+	u_blox_create_frame(ubx_frame, ubx_frame_size, UBX_CLASS_CFG, UBX_CFG_NAV5, NULL, 0);
+}
+
+void u_blox_get_cfg_nav5_set(uint8_t ubx_frame[], uint16_t *ubx_frame_size, enum gnss_mode g_mode,
```
I'm gonna summarize some issues that relate to all functions in this file here:

There is no checking of the `ubx_frame` array size. This should be fixed by making the `ubx_frame_size` an in param, and returning an `int`, which is negative if the `ubx_frame` array is too small, and the size of the final `ubx_frame` on success.
```
int u_blox_create_frame(uint8_t *ubx_frame, uint16_t ubx_frame_size, ...)
```

The `payload` buffer is being allocated on the stack, and is quite large (about 40 bytes), this buffer could either be static, and protected with a spinlock, or the `ubx_frame` buffer could be used directly instead of filling the payload section separately, and then copying it to the `ubx_frame` buffer.

To make using the ubx_frame buffer directly easier, it can be typecast into the following struct
```
struct ubx_frame_buf {
        uint8_t header0;
        uint8_t header1;
        uint8_t message_class;
        uint8_t message_id;
        uint8_t payload_size_low;
        uint8_t payload_size_high;
        uint8_t payload[];
};

struct ubx_frame_buf *buf = (struct ubx_frame_buf *)ubx_frame;
```

bjarki-trackunit commented on drivers/gnss/u_blox_protocol/u_blox_protocol.c:
```diff
@@ -0,0 +1,217 @@
+			     uint16_t t_acc, uint8_t static_hold_thresh, uint8_t dgnss_timeout,
+			     uint8_t cno_thresh_num_svs, uint8_t cno_thresh,
+			     uint16_t static_hold_max_dist, enum utc_standard utc_strd)
+{
```
This function will push 34 + 36 bytes to the stack on entry, then a few more when entering `u_blox_create_frame`. This can be helped by placing data into the `ubx_frame` buffer directly, and by creating a structure with all the arguments passed to this function, then passing a const pointer to said struct to the function.
```
struct u_blox_protocol_nav_5_cfg {
        enum gnss_mode g_mode;
        enum fix_mode f_mode;
        ...
};

struct u_blox_protocol_nav_5_cfg foo_cfg = {
        .g_mode = ,
        .f_mode = ,
};

ublox_get_nav_5_set(buf, sizeof(buf), &foo_cfg);
```
This also makes it easier to pass the correct arguments to the functions since the names are in the struct, rather than having to count the arguments order (there are 17 arguments, many of which can be mixed up without the compiler noticing, like `p_dop` and `h_dop`)

bjarki-andreasen reviewed (changes requested):
I have quite a few requests, I have limited the review to the `u_blox_protocol` related files for now to make it easier to attend to and/or discuss the comments :)

mayankmahajan-nxp commented:
> I have quite a few requests, I have limited the review to the `u_blox_protocol` related files for now to make it easier to attend to and/or discuss the comments :)

Hi @bjarki-trackunit, thank you for the review. I have addressed the comments. Kindly have a look at your convenience.

bjarki-trackunit commented on drivers/gnss/gnss_u_blox_protocol/gnss_u_blox_protocol.h:
```diff
@@ -0,0 +1,276 @@
+#define UBX_CFG_GNSS_FLAG_SGN_CNF_GLONASS_L2	0x10 << UBX_CFG_GNSS_FLAG_SGN_CNF_SHIFT
+
+struct ubx_cfg_gnss_data_config_block {
+	uint8_t gnssId;
```
should be `gnss_id` to match naming convention

bjarki-andreasen reviewed (changes requested):
Looking good, a few more requests, and a question:
* Are the definitions in gnss_u_blox_protocol.h and .c generic, as in, match other ublox gnss than the m10?

mayankmahajan-nxp commented:
> Looking good, a few more requests, and a question:
> * Are the definitions in gnss_u_blox_protocol.h and .c generic, as in, match other ublox gnss than the m10?

Yes, they are generic to ublox gnss devices. For example, ublox Neo M8 (if implemented) would use them as well, without any change.

Also, modem_ubx.c needs some macros in gnss_ublox_protocol.h, which is why I plan to transfer it to the include directory of gnss in the next commit.

mayankmahajan-nxp commented:
Hi @bjarki-trackunit, thank you for the review. I have addressed the comments. Kindly have a look at your convenience.

bjarki-trackunit commented on drivers/gnss/gnss_u_blox_m10.c:
```diff
@@ -0,0 +1,888 @@
+		goto out_1;
+	}
+
+out_1:
```
go to should be named to reflect what it does, see [Linux coding style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#centralized-exiting-of-functions)

I would name `out_1` `free_and_unlock`, and `out_2` `unlock` This goes for the other `out`s as well I think :)

bjarki-trackunit commented on drivers/gnss/gnss_u_blox_m10.c:
```diff
@@ -0,0 +1,888 @@
+
+	key = k_spin_lock(&data->lock);
+
+	UBX_CFG_NAV5_DATA_INIT(frame_data)
```
I don't know what benefit this macro has compared to
```
	struct ubx_cfg_nav5_data frame_data;
	ubx_cfg_nav5_data_default(&frame_data);
```
IMO writing it out is clearer than the macro :) (`struct ubx_cfg_nav5_data frame_data;` should be moved to the top of the function)

bjarki-trackunit commented on drivers/gnss/gnss_u_blox_m10.c:
```diff
@@ -0,0 +1,888 @@
+		.argv_size = ARRAY_SIZE(data->chat_argv),
+		.unsol_matches = unsol_matches,
+		.unsol_matches_size = ARRAY_SIZE(unsol_matches),
+		.process_timeout = K_MSEC(2),
```
The process_timeout no longer exists, please rebase on the latest main and it will be gone :)

bjarki-trackunit commented on include/zephyr/modem/ubx.h:
```diff
@@ -0,0 +1,138 @@
+ * @returns negative errno code if failure
+ */
+int modem_ubx_run_script(struct modem_ubx *ubx, const struct modem_ubx_script *script);
+
```
I believe there is an unclear separation between the responsibility of `ubx.h` and `gnss_u_blox_protocol.h` I understand that the GNSS specific macros and values like `UBX_CFG_GNSS_FLAG_SGN_CNF_GALILEO_E1` are in `gnss_u_blox_protocol.h`, but not `UBX_FRM_MSG_CLASS_IDX` and `UBX_PREAMBLE_SYNC_CHAR_1`. The latter are related to the UBX protocol itself, as opposed to the contents of the payload, which I would understand being split into its own header :)

As I understand it, `ubx.h` should not include `gnss_u_blox_protocol.h`, rather, everything it needs that is currently in `gnss_u_blox_protocol.h` should be moved to `ubx.h`, and `gnss_u_blox_protocol.h` should include `ubx.h` :)

With this, everything related to the UBX frame itself, is in `ubx.h`, checksums, flags, transport etc. while the different message classes, like `CFG(0x06)` are in a "child" header

mayankmahajan-nxp replied:
Understood. Just one question - `modem_ubx` needs the message class and message IDs of UBX-ACK message. And all messages classes and messages IDs are given in `gnss_u_blox_protocol_defines.h`. For now, to resolve the issue, I have added the following macros in `ubx.h`, but it causes some duplication:

```
#define UBX_MSG_CLASS_ACK	0x05
#define UBX_ACK_MSG_ID_ACK	0x01
#define UBX_ACK_MSG_ID_NAK	0x00
```
(These are GNSS specific macros.)

bjarki-trackunit replied:
We have a similar scenario for NMEA and AT messages :) The `modem_chat` module defines two things for each script, what to send, and what to expect to get back. When sending `AT+CEREG`, we expect to get `+CEREG=1,2` back. The `modem_chat` module has no concept of what any AT message is, or what should be returned, just like `modem_ubx` should not need to "understand" what the sent payload is, or what is expected to be returned.

A solution to this issue for `modem_ubx` is to write the script like this:

```
struct modem_ubx_frame {
        /* Class of UBX frame */
        uint8_t class;
        /* ID of UBX frame */
        uint8_t id;
        /* Source of payload for request
         * Destination for payload for response
         */
        uint8_t *payload;
        /* Size of payload to send for request
         * Expected size of payload for response
         */
        uint16_t payload_size;
};

struct modem_ubx_script {
        /* Pointer to request to send */
        struct modem_ubx_frame *request;
        /* Pointer to array of responses in the order they will be received */
        struct modem_ubx_frame *responses;
        /* Size of array of responses */
        uint8_t responses_size;
	uint16_t retry_count;
	k_timeout_t script_timeout;
};
```

This solution splits the encoding of a frame, including header and checksum, to `modem_ubx`, while `gnss_u_blox_protocol_defines.h` sets the classes and ids, formats the payload, and describes which frames are expected to be sent/received in which order :)

bjarki-trackunit replied:
I created this example of how to use it, converting this
```
	UBX_CFG_PRT_SET_DATA_INIT(frame_data)
	frame_data.baudrate = target_baudrate;

	struct modem_ubx_script script;

	ret = ubx_m10_modem_ubx_script_init(dev, &script, data->ubx_frame_buf,
		sizeof(data->ubx_frame_buf), retry, &frame_data, UBX_CLASS_CFG, UBX_CFG_PRT,
		UBX_CFG_PRT_SET_PAYLOAD_SZ);
	if (ret < 0) {
		goto out;
	}

	script.ubx_frame_size = ret;
	ret = ubx_m10_modem_ubx_run_script(dev, &script);
	if (ret < 0) {
		goto out;
	}
```
into
```
	UBX_CFG_PRT_SET_DATA_INIT(frame_data)
	frame_data.baudrate = target_baudrate;

	struct modem_ubx_frame cfg_prt_frame = {
		.class = UBX_CLASS_CFG,
		.id = UBX_CFG_PRT,
		.payload = (const uint8_t *)frame_data,
		.payload = UBX_CFG_PRT_SET_PAYLOAD_SZ,
	};

	struct modem_ubx_script cfg_prt_script = {
		.request = &cfg_prt_frame,
		.responses = NULL,
		.responses_size = 0,
		.retry_count = 3,
		.script_timeout = K_SECONDS(2),
	};

	ret = ubx_m10_modem_ubx_run_script(dev, &script);
	if (ret < 0) {
		goto out;
	}

```

mayankmahajan-nxp replied:
Thank you for answering the question @bjarki-trackunit. Really appreciate your help and effort.
I just pushed my latest changes as you could see below. Will address this comment as well and push the new changes shortly.

bjarki-trackunit replied:
Nice, I will review your latest input shortly

bjarki-andreasen reviewed (changes requested):
Its getting there, but I'm not convinced on the responsibilities of `ubx.h` and `gnss_u_blox_protocol.h`, more on that in the comments :)

mayankmahajan-nxp commented on drivers/gnss/gnss_u_blox_protocol.c:
```diff
@@ -0,0 +1,198 @@
+	data->rate = UBX_CFG_MSG_RATE_DEFAULT;
+}
+
+uint16_t ubx_get_payload_size(uint8_t *ubx_frame)
+{
+	struct ubx_frame_t *frame = (struct ubx_frame_t *) ubx_frame;
+
+	return (uint16_t) ((frame->payload_size_low & 0xFF) | (frame->payload_size_high << 8));
+}
```
Thank you fixing up the code. Will be more cautious moving forward.
Also, I realized I am not using this function anymore - so removed it for now.

mayankmahajan-nxp commented:
Hi @bjarki-trackunit, thank you for the review. I have addressed the comments. Kindly have a look at your convenience.

mayankmahajan-nxp commented:
Hi @bjarki-trackunit, I have addressed this comment - https://github.com/zephyrproject-rtos/zephyr/pull/68350#discussion_r1508675601. Kindly review at your convenience. Thank you.

bjarki-andreasen reviewed (changes requested):
The code is looking really good, these (hopefully) last requests are quite minor, Nice work!

fabiobaltieri commented on drivers/gnss/Kconfig.u_blox_m10:
```diff
@@ -0,0 +1,29 @@
+	select GNSS_U_BLOX_PROTOCOL
+	select UART_USE_RUNTIME_CONFIGURE
+	help
+		Enable U-BLOX M10 GNSS modem driver.
```
nit: indentation is off

fabiobaltieri commented on drivers/gnss/gnss_u_blox_m10.c:
```diff
@@ -0,0 +1,940 @@
+		.message_id = msg_id,
+	};
+
+	(void) ubx_m10_modem_ubx_script_fill(dev);
```
```suggestion
	(void)ubx_m10_modem_ubx_script_fill(dev);
```

same for the rest of the file

fabiobaltieri commented on drivers/gnss/gnss_u_blox_m10.c:
```diff
@@ -0,0 +1,940 @@
+{
+	struct ubx_m10_data *data = dev->data;
+
+	data->script.request = (struct ubx_frame_t *) data->request_buf;
+	data->script.response = (struct ubx_frame_t *) data->response_buf;
+	data->script.match = (struct ubx_frame_t *) data->match_buf;
```
```suggestion
	data->script.request = (struct ubx_frame_t *)data->request_buf;
	data->script.response = (struct ubx_frame_t *)data->response_buf;
	data->script.match = (struct ubx_frame_t *)data->match_buf;
```

fabiobaltieri commented on drivers/gnss/gnss_u_blox_m10.c:
```diff
@@ -0,0 +1,940 @@
+	switch (gnss_system) {
+	case GNSS_SYSTEM_GPS:
+		payload->config_blocks[index].gnss_id = UBX_GNSS_ID_GPS;
+		payload->config_blocks[index].flags = enable |
+			UBX_CFG_GNSS_FLAG_SGN_CNF_GPS_L1C_A;
```
is this correct? boolean | value? guess it's fine it just looks odd to me, maybe add a define for the enable/disable bit and use an intermediate variable or something, just for clarity

mayankmahajan-nxp replied:
Hi @fabiobaltieri, i have addressed this comment in the latest push. Kindly check if the resolution is correct.

fabiobaltieri commented on drivers/gnss/gnss_u_blox_protocol/gnss_u_blox_protocol_defines.h:
```diff
@@ -0,0 +1,258 @@
+};
+
+enum ubx_utc_standard {
+	UBX_UTC_AutoUTC = 0,
```
```suggestion
	UBX_UTC_AUTOUTC = 0,
```

fabiobaltieri commented on drivers/gnss/gnss_u_blox_protocol/gnss_u_blox_protocol.c:
```diff
@@ -0,0 +1,218 @@
+		case UBX_CFG_MSG:
+			payload_size_expected = UBX_CFG_MSG_PAYLOAD_SZ;
+			break;
+		default: return -1;
```
it's in the coding style https://www.kernel.org/doc/html/v4.10/process/coding-style.html#indentation

`Don’t put multiple statements on a single line unless you have something to hide`

fabiobaltieri commented on subsys/modem/modem_ubx.c:
```diff
@@ -0,0 +1,287 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(modem_ubx, CONFIG_MODEM_MODULES_LOG_LEVEL);
+
+#define MODEM_UBX_STATE_ATTACHED_BIT		(0)
```
```suggestion
#define MODEM_UBX_STATE_ATTACHED_BIT		0
```

mayankmahajan-nxp commented on drivers/gnss/gnss_u_blox_protocol/gnss_u_blox_protocol.c:
```diff
@@ -0,0 +1,218 @@
+	frame->payload_and_checksum[payload_size + 1] = ckB;
+
+	return ubx_frame_len;
+}
```
Hi @bjarki-trackunit, I have moved this to `modem_ubx.c`, but please note that payload size couldn't be validated in `modem_ubx.c` because details regarding gnss ublox messages is concern of `gnss_u_blox_protocol.c`. Hence, I have created a function in `ubx_create_and_validate_frame` in `gnss_u_blox_protocol.c`  that validates the payload size then calls `modem_ubx_create_frame` in `modem_ubx.c` to create the frame. Please check if this is correct.

fabiobaltieri commented on drivers/gnss/CMakeLists.txt:
```diff
@@ -9,3 +9,6 @@ zephyr_library_sources_ifdef(CONFIG_GNSS_NMEA0183 gnss_nmea0183.c)
 zephyr_library_sources_ifdef(CONFIG_GNSS_NMEA_GENERIC gnss_nmea_generic.c)
 zephyr_library_sources_ifdef(CONFIG_GNSS_QUECTEL_LCX6G gnss_quectel_lcx6g.c)
+zephyr_library_sources_ifdef(CONFIG_GNSS_U_BLOX_M10 gnss_u_blox_m10.c)
+zephyr_library_sources_ifdef(CONFIG_GNSS_U_BLOX_PROTOCOL
+			     gnss_u_blox_protocol/gnss_u_blox_protocol.c)
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_GNSS_U_BLOX_PROTOCOL gnss_u_blox_protocol/gnss_u_blox_protocol.c)
```

mayankmahajan-nxp commented:
Hi @bjarki-trackunit and @fabiobaltieri, thank you for the review. I have addressed the latest comments. Kindly have a look at your convenience.

fabiobaltieri commented on drivers/gnss/gnss_u_blox_m10.c:
```diff
@@ -0,0 +1,1058 @@
+/* The datasheet of the device doesn't specify boot time. But 1 sec helped significantly. */
+#define UBX_M10_BOOT_TIME_MS		1000
+
+enum MODEM_MODULE {
```
```suggestion
enum modem_module {
```

if you really want an enum here, though to me it looks like you could just replace this with a boolean and everything would be come a tiny bit simpler, do you forsee this getting more values in the future? The whole `ubx_m10_modem_module_switch` just looks really really weird right now

mayankmahajan-nxp replied:
@fabiobaltieri We don't foresee more modem modules being used in this driver. We could utilize a boolean macro, but i thought enum improves readability. I would prefer an enum here, but if you want i will change this as you said.

fabiobaltieri replied:
Ok, I think it should be changed, I mean this would maybe make sense if you wanted to do stuff like `ubx_m10_modem_module_switch(MODEM_MODULE_CHAT, MODEM_MODULE_CHAT);`, it's just odd to have an enum for what is really a boolean. @bjarki-trackunit?

mayankmahajan-nxp replied:
Yes @fabiobaltieri, I understand your point now. Passing two arguments to `ubx_m10_modem_module_switch` isn't required. Would the following work? 

```
/**
 * @brief Changes modem module (chat or ubx) attached to the uart pipe.
 * @param dev Dev instance
 * @param change_from_to 0 for changing from "chat" to "ubx", 1 for changing from "ubx" to "chat"
 * @returns 0 if successful
 * @returns negative errno code if failure
 */
static int ubx_m10_modem_module_change(const struct device *dev, bool change_from_to)
{
	struct ubx_m10_data *data = dev->data;
	int ret;

	if (change_from_to == 0) {
		modem_chat_release(&data->chat);
		ret = modem_ubx_attach(&data->ubx, data->uart_pipe);
	} else { /* change_from_to == 1 */
		modem_ubx_release(&data->ubx);
		ret = modem_chat_attach(&data->chat, data->uart_pipe);
	}

	if (ret < 0) {
		(void)modem_pipe_close(data->uart_pipe);
	}

	return ret;
}
```

Ps. I would prefer an enum here (or a couple of macros). Because without these, while calling `ubx_m10_modem_module_change`, we have to pass a literal. For now I have pushed the changes as per your comments.
```
enum modem_module_change {
	MODEM_MODULE_CHAT_TO_UBX = 0,
	MODEM_MODULE_UBX_TO_CHAT = 1,
};
```

bjarki-trackunit replied:
the function effectively attaches either chat or ubx pipe, modem_pipe_attach will automatically release the currently attached pipe, so simply attaching the required pipe should be clear and concise, without requiring a helper function at all :)

fabiobaltieri commented on subsys/modem/Kconfig:
```diff
@@ -54,6 +54,21 @@ config MODEM_PPP_NET_BUF_FRAG_SIZE
+	select RING_BUFFER
+	select MODEM_PIPE
+	help
+		Enable Modem U-BLOX module.
```
```suggestion
	  Enable Modem U-BLOX module.
```

mayankmahajan-nxp commented:
Hi @fabiobaltieri, thank you for the review. I have addressed the comments.

fabiobaltieri commented on drivers/gnss/gnss_u_blox_protocol/gnss_u_blox_protocol.h:
```diff
@@ -0,0 +1,251 @@
+#define UBX_CFG_NAK_PAYLOAD_SZ			2
+#define UBX_CFG_RATE_PAYLOAD_SZ			6
+#define UBX_CFG_PRT_POLL_PAYLOAD_SZ		1
+#define UBX_CFG_PRT_POLL_FRM_SZ			UBX_FRM_SZ_WO_PAYLOAD + UBX_CFG_PRT_POLL_PAYLOAD_SZ
```
this and the other defines with operations should be parenthesized, just to be sure there's no hidden operation priority errors

fabiobaltieri commented on drivers/gnss/gnss_u_blox_protocol/gnss_u_blox_protocol.h:
```diff
@@ -0,0 +1,251 @@
+#define UBX_CFG_PRT_PORT_MODE_STOP_BITS_1		0U
+#define UBX_CFG_PRT_PORT_MODE_STOP_BITS_1_HALF		BIT(12)
+#define UBX_CFG_PRT_PORT_MODE_STOP_BITS_2		BIT(13)
+#define UBX_CFG_PRT_PORT_MODE_STOP_BITS_HALF		BIT(12) | BIT(13)
```
same for this one, just the ones with multiple operands

fabiobaltieri commented on drivers/gnss/gnss_u_blox_protocol/gnss_u_blox_protocol.h:
```diff
@@ -0,0 +1,251 @@
+#define UBX_CFG_GNSS_FLAG_DISABLE		0U
+#define UBX_CFG_GNSS_FLAG_SGN_CNF_SHIFT		16
+/* When gnss_id is 0 (GPS) */
+#define UBX_CFG_GNSS_FLAG_SGN_CNF_GPS_L1C_A	0x01 << UBX_CFG_GNSS_FLAG_SGN_CNF_SHIFT
```
same

fabiobaltieri reviewed (commented):
Heya, just a comment on definitions, sorry I think I asked you to drop the parenthesis to the literal ones before, I meant to ask that only for definitions with a single literal, the others should have them otherwise weird things can happen depending on how they are used.

bjarki-andreasen reviewed (approved):
Nice work!

JordanYates commented:
I know I'm a few months late here, but I'm curious how this works given that the Ublox protocol used on the M10 chipsets no longer support the old dedicated `CFG-*` messages? Specifically configuration like `CFG-RATE` needs to be supplied through the new `CFG-VALSET` messages.

For example:
![image](https://github.com/user-attachments/assets/3d70bb83-56bb-4063-9517-71b6b347ef85)

mayankmahajan-nxp commented:
Hi @JordanYates, thank you for the comment. I will look into this and get back to you soon.
